librz/core/cbin.c

```diff
@@ -2305,6 +2305,748 @@ static const char *bin_reloc_type_name(RzBinReloc *reloc) {
+	pj_ks(pj, "perm", perms);
+	if (!section->is_segment) {
+		char *section_type = rz_bin_section_type_to_string(core->bin, section->type);
+		if (section_type) {
+			pj_ks(pj, "type", section_type);
+		}
+		free(section_type);
+	}
+	if (!section->is_segment) {
+		RzList *flags = rz_bin_section_flag_to_list(core->bin, section->flags);
+		if (!rz_list_empty(flags)) {
+			RzListIter *it;
+			char *pos;
+			pj_ka(pj, "flags");
+			rz_list_foreach (flags, it, pos) {
+				pj_s(pj, pos);
+			}
+			pj_end(pj);
+		}
+		rz_list_free(flags);
+	}
+	pj_kN(pj, "paddr", section->paddr);
+	pj_kN(pj, "vaddr", addr);
+	if (section->align) {
+		pj_kN(pj, "align", section->align);
+	}
+	if (hashes && section->size > 0) {
+		ut8 *data = malloc(section->size);
+		if (data) {
+			ut32 datalen = section->size;
+			RzListIter *iter;
+			char *hashname;
+
+			rz_io_pread_at(core->io, section->paddr, data, datalen);
+
+			rz_list_foreach (hashes, iter, hashname) {
+				char *chkstr = rz_msg_digest_calculate_small_block_string(hashname, data, datalen, NULL, false);
+				if (!chkstr) {
+					continue;
+				}
+				pj_ks(pj, hashname, chkstr);
+				free(chkstr);
+			}
+			free(data);
+		}
+	}
+	pj_end(pj);
+}
+
+static void sections_print_table(RzCore *core, RzTable *t, RzBinObject *o, RzBinSection *section, RzList *hashes) {
+	ut64 addr = get_section_addr(core, o, section);
+	char perms[5];
+	section_perms_str(perms, section->perm);
+
+	RzList *row_list = rz_list_newf(free);
+	if (!row_list) {
+		return;
+	}
+	char *section_type = NULL;
+	if (!section->is_segment) {
+		section_type = rz_bin_section_type_to_string(core->bin, section->type);
+	}
+	char *section_flags_str = NULL;
+	if (!section->is_segment) {
+		RzList *section_flags = rz_bin_section_flag_to_list(core->bin, section->flags);
+		section_flags_str = rz_str_list_join(section_flags, ",");
+		rz_list_free(section_flags);
+	}
+
+	rz_table_add_rowf(t, "XxXxss", section->paddr, section->size, addr, section->vsize, perms, section->name);
+	if (!section->is_segment) {
+		rz_table_add_row_columnsf(t, "ss", section_type, section_flags_str);
+	}
+	rz_table_add_row_columnsf(t, "x", section->align);
+	if (hashes && section->size > 0) {
+		ut8 *data = malloc(section->size);
+		if (data) {
+			ut32 datalen = section->size;
+			RzListIter *iter;
+			char *hashname;
+
+			rz_io_pread_at(core->io, section->paddr, data, datalen);
+
+			rz_list_foreach (hashes, iter, hashname) {
+				const RzMsgDigestPlugin *msg_plugin = rz_msg_digest_plugin_by_name(hashname);
+				if (!msg_plugin) {
+					continue;
+				}
+				char *chkstr = rz_msg_digest_calculate_small_block_string(hashname, data, datalen, NULL, false);
+				if (!chkstr) {
+					rz_table_add_row_columnsf(t, "s", NULL);
+					continue;
+				}
+				rz_table_add_row_columnsf(t, "s", chkstr);
+				free(chkstr);
+			}
+			free(data);
+		}
+	}
+}
+
+RZ_IPI void rz_core_bin_sections_print(RzCore *core, RzCmdStateOutput *state, RzList *hashes) {
+	RzBinFile *bf = rz_bin_cur(core->bin);
```

Comment:
Maybe a quick exit if `hashes` is empty?

Reply:
no if there are no hashes we just skip them, but you still print sections/segments.